### PR TITLE
Optimize buffer utility functions

### PR DESCRIPTION
**Pull request by jakirkham**
Rewrites buffer utility functions to get more optimizations out of Cython. This includes typing variables where possible, leveraging `for`-loops that are easier for Cython to lower to C, making use of Cython `memoryview`s, and some code simplification (where possible).

cc @pentschev @madsbk @quasiben

### CODE REVIEW

**Comment by pentschev** [comment]
@jakirkham I'm getting some errors when I try the send/recv debug test:

```
$ UCX_TLS=tcp,sockcm,cuda_copy UCX_SOC
KADDR_TLS_PRIORITY=sockcm python recv.py
'CUDA_VISIBLE_DEVICES'
ITER:  0
IndexError: Out of bounds on buffer access (axis 0)
Exception ignored in: 'ucp._libs.utils.get_buffer_nbytes'
IndexError: Out of bounds on buffer access (axis 0)
Traceback (most recent call last):
  File "recv.py", line 140, in <module>
    test_send_recv_cu(cupy)
  File "recv.py", line 113, in test_send_recv_cu
    client(env2, port, func)
  File "recv.py", line 58, in client
    asyncio.get_event_loop().run_until_complete(read())
  File "/datasets/pentschev/miniconda3/envs/rn-102-0.15.0b200615/lib/python3.7/asyncio/base_events.py", line 583, in run_until_complete
    return future.result()
  File "recv.py", line 47, in read
    frames, msg = await recv(ep)
  File "/datasets/pentschev/src/ucx-py/debug-tests/utils.py", line 104, in recv
    await ep.recv(frame)
  File "/datasets/pentschev/miniconda3/envs/rn-102-0.15.0b200615/lib/python3.7/site-packages/ucp/core.py", line 604, in recv
    ret = await ucx_api.tag_recv(self._ep, buffer, nbytes, tag, log_msg=log)
ucp.exceptions.UCXMsgTruncated: Comm Error "[Recv #005] ep: 0x7fcf28038048, tag: 0x68f5dfcc8a2d7c57, nbytes: 0, type: <class 'rmm._lib.device_buffer.DeviceBuffer'>": length mismatch: 472392 (got) != 0 (expected)
```

I'm guessing this has to do with some of the logic change in `get_buffer_nbytes`, but I'm not sure if this is ready to test or you're still working on it.

**Comment by jakirkham** [comment]
Yeah there are not really tests of this functionality, which makes it hard to tell what would be expected of these functions. Think we should get some tests in first and then revisit.

**Comment by jakirkham** [comment]
I've added some tests in PR ( https://github.com/rapidsai/ucx-py/pull/549 ).

File: ucp/_libs/utils.pyx [hunk 1]
```diff
@@ -42,41 +42,52 @@ def get_buffer_data(buffer, check_writable=False):
...
+    cdef int64_t[::1] strides
+    cdef size_t nbytes
     if iface is not None:
         import numpy
-        itemsize = int(numpy.dtype(iface['typestr']).itemsize)
+        itemsize = numpy.dtype(iface["typestr"]).itemsize
         # Making sure that the elements in shape is integers
-        shape = [int(s) for s in iface['shape']]
-        nbytes = reduce(operator.mul, shape, 1) * itemsize
+        shape = numpy.array(iface["shape"], dtype=numpy.uint64)
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:73]: What is `iface["shape"]` here? A tuple? In Numba previously @shwina found that allocating a numpy array to do a cumulative product of a tuple is much more expensive then just doing an `itertools.reduce` for example. We should likely be even more efficient in using the Python C-API here if we wanted and iterating over the tuple via that.

**Comment by jakirkham** [comment]
I wouldn't worry about reviewing this yet. We seem to still be getting failures atm. Would be good to get some tests in first ( https://github.com/rapidsai/ucx-py/pull/549 ) so we can better understand the origin of the failures.

File: ucp/_libs/utils.pyx (hunk 1 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx:73]: Yeah wasn't planning to stick with NumPy. In any event there are a few ways of coercing something like a `tuple` to a C-style array. Of them just using the builtin `array` should be fine for small arrays (as seen below). This was run on a DGX-1.

```python
In [1]: from array import array 
   ...: import struct 
   ...:  
   ...: import numpy                                                            

In [2]: t = (2, 3, 4)                                                           

In [3]: %timeit numpy.array(t)                                                  
1.17 µs ± 2.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit array("l", t)                                                   
243 ns ± 0.105 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit struct.pack(f"{len(t)}l", *t)                                   
368 ns ± 1.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

File: ucp/_libs/utils.pyx [hunk 2]
```diff
@@ -1,38 +1,40 @@
...
-from libc.stdint cimport uintptr_t
+from libc.stdint cimport int64_t, uint64_t, uintptr_t
+
+from array import array
 
 from ..exceptions import UCXCloseError, UCXError
 
 
-def get_buffer_data(buffer, check_writable=False):
+cpdef uintptr_t get_buffer_data(buffer, bint check_writable=False) except *:
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:13]:
```suggestion
cpdef uintptr_t get_buffer_data(object buffer, bint check_writable=False) except *:
```

File: ucp/_libs/utils.pyx (hunk 1 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx]: This now uses `array`. So marking this as resolved.

File: ucp/_libs/utils.pyx [hunk 3]
```diff
@@ -43,44 +45,55 @@ def get_buffer_data(buffer, check_writable=False):
     return data_ptr
 
 
-def get_buffer_nbytes(buffer, check_min_size, cuda_support):
+cpdef size_t get_buffer_nbytes(buffer, check_min_size, bint cuda_support) except *:
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:48]: Not sure on the typing of `check_min_size` here:
```suggestion
cpdef size_t get_buffer_nbytes(object buffer, object check_min_size, bint cuda_support) except *:
```

File: ucp/_libs/utils.pyx [hunk 4]
```diff
@@ -1,38 +1,40 @@
...
 
 
-def get_buffer_data(buffer, check_writable=False):
+cpdef uintptr_t get_buffer_data(buffer, bint check_writable=False) except *:
     """
     Returns data pointer of the buffer. Raising ValueError if the buffer
     is read only and check_writable=True is set.
     """
-    iface = None
+    cdef dict iface = None
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:18]: I didn't know this is allowed, neat!

File: ucp/_libs/utils.pyx [hunk 5]
```diff
@@ -43,44 +45,55 @@ def get_buffer_data(buffer, check_writable=False):
...
         import numpy
-        itemsize = int(numpy.dtype(iface['typestr']).itemsize)
+        itemsize = numpy.dtype(iface["typestr"]).itemsize
         # Making sure that the elements in shape is integers
-        shape = [int(s) for s in iface['shape']]
-        nbytes = reduce(operator.mul, shape, 1) * itemsize
+        shape = array("L", iface["shape"])
+        nbytes = itemsize
+        for i in range(len(shape)):
+            nbytes *= shape[i]
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:82]: Is constructing an array from the tuple and then iterating over the array actually cheaper than just iterating over the tuple itself?

**Comment by jakirkham** [comment]
We are seeing a test failure due to a recent change in Dask ( https://github.com/dask/distributed/pull/4019 ). Submitted PR ( https://github.com/dask/distributed/pull/4036 ) to fix that.

File: ucp/_libs/utils.pyx (hunk 2 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx:13]: Cython already defaults to `object` if a type is specified. So we are ok here.

File: ucp/_libs/utils.pyx (hunk 3 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx:48]: Cython already defaults to `object` if a type is specified. So we are ok here.

Yeah `check_min_size` could be several things. So `object` is probably as close as we can get.

**Comment by jakirkham** [comment]
rerun tests

File: ucp/_libs/utils.pyx [hunk 6]
```diff
@@ -42,41 +42,52 @@ def get_buffer_data(buffer, check_writable=False):
...
+            if len(shape) > 0:
+                if len(strides) != len(shape):
+                    raise ValueError(
+                        "The length of shape and strides must be equal"
+                    )
+                s = <int64_t>itemsize
+                for i from len(shape) >= i >= 0 by 1:
+                    if s != strides[i]:
+                        raise ValueError("Array must be contiguous")
+                    s *= <int64_t>shape[i]
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply ucp/_libs/utils.pyx]: This was done in recent commits.

File: ucp/_libs/utils.pyx [hunk 7]
```diff
@@ -16,23 +18,17 @@ def get_buffer_data(buffer, check_writable=False):
...
     """
-    iface = None
-    if hasattr(buffer, "__cuda_array_interface__"):
-        iface = buffer.__cuda_array_interface__
-    elif hasattr(buffer, "__array_interface__"):
-        iface = buffer.__array_interface__
+
+    iface = getattr(buffer, "__array_interface__", None)
+    if iface is None:
+        iface = getattr(buffer, "__cuda_array_interface__", None)
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:22]: Can we swap these to say the extra `getattr` call in the case of GPU memory? 😄

File: ucp/_libs/utils.pyx [hunk 8]
```diff
@@ -16,23 +18,17 @@ def get_buffer_data(buffer, check_writable=False):
...
         mview = memoryview(buffer)
-        data_ptr = int(<uintptr_t>PyMemoryView_GET_BUFFER(mview).buf)
-        data_readonly = mview.readonly
-
-    # Workaround for numba giving None, rather than an 0.
-    # https://github.com/cupy/cupy/issues/2104 for more info.
-    if data_ptr is None:
-        data_ptr = 0
+        data_ptr = <uintptr_t>PyMemoryView_GET_BUFFER(mview).buf
+        data_readonly = <bint>PyMemoryView_GET_BUFFER(mview).readonly
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:29]: Can store an intermediate of the result of `PyMemoryView_GET_BUFFER(mview)` here to save a call.

File: ucp/_libs/utils.pyx [hunk 9]
```diff
@@ -49,38 +45,40 @@ def get_buffer_nbytes(buffer, check_min_size, cuda_support):
...
-                  "variable and that the ucx-proc=*=gpu package is " \
-                  "installed. See " \
-                  "https://ucx-py.readthedocs.io/en/latest/install.html for " \
-                  "more information."
-            raise ValueError(msg)
-    elif hasattr(buffer, "__array_interface__"):
-        iface = buffer.__array_interface__
+    iface = getattr(buffer, "__array_interface__", None)
+    if iface is None:
+        iface = getattr(buffer, "__cuda_array_interface__", None)
```
Comment (kkraus14) [inline ucp/_libs/utils.pyx:48]: Similarly can we swap these?

Comment (jakirkham) [reply ucp/_libs/utils.pyx]: We need to check whether we have a CUDA object and whether those are allowed in this case below.

File: ucp/_libs/utils.pyx (hunk 8 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx:28]: [`PyMemoryView_GET_BUFFER`]( https://github.com/python/cpython/blob/d905df766c367c350f20c46ccd99d4da19ed57d8/Include/memoryobject.h#L18 ) is just a macro that access the `Py_buffer` object owned by the `memoryview`.

File: ucp/_libs/utils.pyx [hunk 10]
```diff
@@ -1,38 +1,31 @@
...
+        data_ptr, data_readonly = iface["data"]
     else:
         mview = memoryview(buffer)
-        data_ptr = int(<uintptr_t>PyMemoryView_GET_BUFFER(mview).buf)
-        data_readonly = mview.readonly
-
-    # Workaround for numba giving None, rather than an 0.
-    # https://github.com/cupy/cupy/issues/2104 for more info.
-    if data_ptr is None:
-        data_ptr = 0
```
Comment (jakirkham) [inline ucp/_libs/utils.pyx:35]: This was fixed in Numba PR ( https://github.com/numba/numba/pull/4609 ), which was included in 0.46.0+ released last October. We require [`numba >=0.49.0`]( https://github.com/rapidsai/integration/blob/5a6b620e70ed774b202ad91868adcf325d23df11/conda/recipes/versions.yaml#L96-L97 ) currently and likely will need to bump that before RAPIDS 0.15 is released. So I think we are safe to drop this.

File: ucp/_libs/utils.pyx [hunk 11]
```diff
@@ -43,51 +37,62 @@ def get_buffer_data(buffer, check_writable=False):
...
-                  "variable and that the ucx-proc=*=gpu package is " \
-                  "installed. See " \
-                  "https://ucx-py.readthedocs.io/en/latest/install.html for " \
-                  "more information."
-            raise ValueError(msg)
-    elif hasattr(buffer, "__array_interface__"):
-        iface = buffer.__array_interface__
+    cdef dict iface = getattr(buffer, "__array_interface__", None)
+    if iface is None:
+        iface = getattr(buffer, "__cuda_array_interface__", None)
```
Comment (madsbk) [inline ucp/_libs/utils.pyx:50]: Ditto

File: ucp/_libs/utils.pyx [hunk 12]
```diff
@@ -1,38 +1,32 @@
...
     """
-    iface = None
-    if hasattr(buffer, "__cuda_array_interface__"):
-        iface = buffer.__cuda_array_interface__
-    elif hasattr(buffer, "__array_interface__"):
-        iface = buffer.__array_interface__
 
+    cdef dict iface = getattr(buffer, "__array_interface__", None)
+    if iface is None:
+        iface = getattr(buffer, "__cuda_array_interface__", None)
```
Comment (madsbk) [inline ucp/_libs/utils.pyx:20]: Don't we want `"__cuda_array_interface__"` to take precedence over `"__array_interface__"`?

File: ucp/_libs/utils.pyx (hunk 11 above)
Comment (madsbk) [inline ucp/_libs/utils.pyx:50]: Ditto

Comment (jakirkham) [reply ucp/_libs/utils.pyx:50]: Please see [this comment]( https://github.com/rapidsai/ucx-py/pull/546#discussion_r468342638 ).

**Comment by jakirkham** [comment]
Went back through and profiled things carefully. Noted that `hasattr` was costing us (more details in the OP of PR https://github.com/rapidsai/ucx-py/pull/573 since included here). So changed the code to use `getattr` instead, which helped quite a bit. Then made some other similar changes just focusing on the Python code.

After that went back and readded Cython types to the Python variables and converted comprehensions to `for`-loops. Though have skipped dealing with arrays. Arrays are a bit slower, but maybe not noticeable given the overall runtime. Maybe they can be revisited in another PR, but would prefer to skip them for now.

With all of these changes, went through and timed the handling of an N-D array (including the case where the C-contiguous check would fail). Have included the times before and after below. In short, all cases are more than 2x faster. This doesn't sound like much, but these functions are already running `~1µs`.

<details>
<summary>Runtime <b>before</b> of <code>get_buffer_data</code> and <code>get_buffer_nbytes</code>:</summary>

<br>

```python
In [1]: import cupy 
   ...: import numpy 
   ...:  
   ...: from ucp._libs.utils import get_buffer_data, get_buffer_nbytes                                

In [2]: def run(func, *args, **kwargs): 
   ...:     try: 
   ...:         func(*args, **kwargs) 
   ...:     except Exception: 
   ...:         pass 
   ...:                                                                                               

In [3]: s = (10, 11, 12, 13) 
   ...: a = numpy.arange(numpy.prod(s)).reshape(s)                                                    

In [4]: %timeit get_buffer_data(a)                                                                    
2.72 µs ± 8.47 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit get_buffer_nbytes(a, check_min_size=None, cuda_support=True)                          
3.7 µs ± 11.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit run(get_buffer_nbytes, a.T, check_min_size=None, cuda_support=True)                   
4.69 µs ± 18.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

</details>

<details>
<summary>Runtime <b>after</b> of <code>get_buffer_data</code> and <code>get_buffer_nbytes</code></summary>

<br>

```python
In [1]: import cupy 
   ...: import numpy 
   ...:  
   ...: from ucp._libs.utils import get_buffer_data, get_buffer_nbytes                                

In [2]: def run(func, *args, **kwargs): 
   ...:     try: 
   ...:         func(*args, **kwargs) 
   ...:     except Exception: 
   ...:         pass 
   ...:                                                                                               

In [3]: s = (10, 11, 12, 13) 
   ...: a = numpy.arange(numpy.prod(s)).reshape(s)                                                    

In [4]: %timeit get_buffer_data(a)                                                                    
1.18 µs ± 9.89 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit get_buffer_nbytes(a, check_min_size=None, cuda_support=True)                          
1.66 µs ± 9.66 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [6]: %timeit run(get_buffer_nbytes, a.T, check_min_size=None, cuda_support=True)                   
2.33 µs ± 8.45 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

</details>

Additionally we fast path objects through `get_buffer_nbytes` in particular if their `strides is None`. Currently CuPy, Numba, and NumPy already set `strides` to `None` if the data is C-contiguous. Plus serialization through Dask turns any non-contiguous array into a 1-D contiguous array. So any array that Dask gives us is sent through the fast path.

This was not true of cuDF `Buffer`s or RMM `DeviceBuffer`s for historical reasons (CuPy and Numba compatibility issues). Though these issues no longer apply, so have submitted PR ( https://github.com/rapidsai/cudf/pull/5917 ) and PR ( https://github.com/rapidsai/rmm/pull/477 ), which will fast path these objects in the future.

In short, I think collectively we've probably gotten as much out of this exercise as we can for the moment. We might be able to make further improvements by refactoring (contiguity check can be separate from size check). It's also possible a lightweight extension type could help us find the right balance between extracting the array interface and letting us query different common things easily as needed. Though this can probably be saved for a future PR.

File: ucp/_libs/utils.pyx (hunk 9 above)
Comment (pentschev) [reply ucp/_libs/utils.pyx]: But what happens if an array includes both `__array_interface__` and `__cuda_array_interface__`? It seems that the latter would never be used, though I don't know if there are any arrays that implement both.

Comment (pentschev) [reply ucp/_libs/utils.pyx]: What would happen if an array implements both? It seems that then we would never try `__cuda_array_interface__`, though I don't know if this is valid and if some array type implements both.

Comment (pentschev) [reply ucp/_libs/utils.pyx]: Sorry, I thought I thought I had lost my first comment, ignore the second which is basically the same.

Comment (jakirkham) [reply ucp/_libs/utils.pyx]: Do you know of a case where this would happen? TBH I don’t know of any libraries outside of NumPy that use `__array_interface__`. Most libraries rely on the Python Buffer Protocol (typically used via `memoryview`s).

Comment (jakirkham) [reply ucp/_libs/utils.pyx]: As NumPy also supports the Python Buffer Protocol, it might be worth dropping the `__array_interface__` handling here in favor of just routing that through a `memoryview` also.

Comment (pentschev) [reply ucp/_libs/utils.pyx]: If the check for `__array_interface__` is exclusive to find whether it's a NumPy array, shouldn't we make it more clear by just checking `isinstance(buffer, np.ndarray)`? The `memoryview` idea makes sense as well, but it remains the question as to what order we should use to do the checking, `__cuda_array_interface__` or `memoryview` first.

Comment (jakirkham) [reply ucp/_libs/utils.pyx]: If we do `memoryview`, we just drop the `__array_interface__` check. `memoryview` is already checked after `__cuda_array_interface__`.

**Comment by jakirkham** [comment]
Switched to using `memoryview` for NumPy. This lets us get rid of another attribute check (already shown to be expensive). Plus we now check `__cuda_array_interface__` first (as the `memoryview` case was a fallback).

File: ucp/_libs/utils.pyx (hunk 12 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx]: `"__cuda_array_interface__"` now takes priority.

File: ucp/_libs/utils.pyx (hunk 11 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx]: `"__cuda_array_interface__"` now takes priority.

File: ucp/_libs/utils.pyx (hunk 9 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx]: Went ahead and switched NumPy to `memoryview` case in recent commits. So marking this resolved. Though if there are additional thoughts please feel free to start a new thread.

File: ucp/_libs/utils.pyx (hunk 7 above)
Comment (jakirkham) [reply ucp/_libs/utils.pyx]: `"__cuda_array_interface__"` is now higher priority.

**Comment by jakirkham** [comment]
Have now also exposed the utility functions directly in Cython to leverage faster calling between Cython code. In Python we are still able to import the functions and use them from Python. So the Python code paths are unaffected.

**Comment by quasiben** [comment]
I know I'm late but this is a really nice performance improvement!
